### ISSUE TITLE
If a reducer throws an exception the store no longer dispatches actions

### ISSUE THREAD

**Issue by kevin--** [arximboldi/lager#156]
If we modify the counter example slightly, we can demonstrate this behavior.
`counter.hpp`:
```
inline model update(model c, action action)
{
    return std::visit(lager::visitor{
                          [&](increment_action) { return model{c.value + 1}; },
                          [&](decrement_action) { return model{c.value - 1}; },
                          [&](reset_action a) {
                               throw std::runtime_error("Cannot reset!");
                               return model{a.new_value};
                          },
                      },
                      action);
}
```

`main.cpp`:
```
int main()
{
    auto store = lager::make_store<counter::action>(
        counter::model{}, lager::with_manual_event_loop{});
    watch(store, draw);

    auto event = char{};
    while (std::cin >> event) {
        try {
            if (auto act = intent(event))
                store.dispatch(*act);
        }
        catch( const std::exception& ex ) {
            std::cout << "sorry couldn't process command: " << ex.what() << "\n";
        }
    }
}
```

if the user inputs "." for Reset, then the application doesn't crash, however further commands of "+" and "-" have no effect, and the reducer root function `update` is not called at all.

This is a little problematic for recovering from non fatal conditions. I would expect that if an exception is thrown, the resulting remains unchanged and the store is in OK condition to continue operating on future commands.

Is this a bug or is this behavior by design? I am a bit out of date, currently using `5feb9c71b15bf6e41d5245496b1aa4eb580c5025`

**Comment by arximboldi** [comment]
That's a bug in `with_manual_event_loop`. I will fix this!

Meanwhile, I use this [store enhancer](https://sinusoid.es/lager/modularity.html#enhancer-interface) to catch exceptions inside a reducer:
```c++
constexpr auto with_exception_logging = [](auto&& next) {
    return [next](auto action,
                  auto&& model,
                  auto&& reducer,
                  auto&& loop,
                  auto&& deps,
                  auto&& tags) {
        return next(
            action,
            LAGER_FWD(model),
            [reducer](auto&& m, auto&& a) -> decltype(reducer(m, a)) {
                try {
                    auto [r, e] = reducer(m, LAGER_FWD(a));
                    return {std::move(r), [e = std::move(e)](auto&& ctx) {
                                try {
                                    e(LAGER_FWD(ctx));
                                } catch (const std::exception& err) {
                                    SPDLOG_ERROR(
                                        "uncaught exception in effect: {}",
                                        err.what());
                                }
                            }};
                } catch (const std::exception& err) {
                    SPDLOG_ERROR("uncaught exception in reducer: {}", err.what());
                    return LAGER_FWD(m);
                }
            },
            LAGER_FWD(loop),
            LAGER_FWD(deps),
            LAGER_FWD(tags));
    };
};
```
This prevents the exception from leaving the reducer. As a downside, it does logging inside the reducer, which is a side-effect, but maybe a tolerable one...

**Comment by kevin--** [comment]
Ok very good to know! The ideal use case is to let the exception leave the reducer, so that we can serialize it and send it to our GUI for the user to read. Its a more natural flow given our current code structure:

for example
```
void deserializeFromGui( msg ) {
    try {
        auto concreteAction = deserialize( msg );
        store.dispatch(concreteAction);
    } catch( const std::exception& ex ) {
        // reply to GUI with error info
    }
}
```

**Comment by arximboldi** [comment]
Hmm interesting. I normally use store loops that allow asynchronous dispatching, so the code you are writing, catching around the `dispatch()` would normally not work. In fact, it only works for `manual_event_loop`, which I never considered very useful for a serious application... but then again, I am happy that you are finding that useful and that this pattern works for you. I will fix the bug ASAP.

**Comment by kevin--** [comment]
much appreciated, indeed we are using `manual_event_loop` and I never considered another approach 😅 Our application framework is entirely custom, so none of the provided event loop implementations were relevant. Though perhaps there is some guidance on the real difference between `queue` and `manual`?

**Comment by kevin--** [comment]
Also in your example with the store enhancer, is this perhaps a better, more generic approach decoupled from the event loop you choose? Because then we can control the business logic of what happens to the model? or do I misunderstand.

**Comment by arximboldi** [comment]
Hi @kevin--, I just submitted a fix for the issue in #157.

However, what you said is true, I would recommend handling exceptions with an enhancer of the form above (or directly in your reducer).

There is still some trouble with letting exception get back into the event loop, as the safety of such operation really depends on how the event loop manages it. In most cases the event loop is asynchronous, and offers us very little opportunity to hook our custom handler in an elegant way.

With a custom event handler directly in the reducer you can really control how the exception is handled and processed, and avoid potential trouble caused by the exception reaching the event loop.

Also, of course I am just guessing here since I see very little in your example, but by using `manual_event_loop`, I feel that you are maybe missing on some of the potential that Lager can provide in your application.

**Comment by kevin--** [comment]
fantastic 🙏 -- we can adopt your fix in the short term and reevaluate some other error handling approaches as discussed.
